execstats: add latch wait time into cumulative contention time

DrewKimball opened the pull request:
#### sql: add setting to enable registering latch wait contention events

This commit adds a session setting (default off) that will cause contention
events emitted by the latch table to be included in the `*_contention_events`
virtual tables. This behavior is hidden behind a setting because latch wait
events are more common than events emitted by the lock table. Tracking them
in the virtual tables could increase system load and flood the statement
insights with uninteresting events. For now, this setting does nothing, but
 the following commit will use it.

Informs #103713

Release note: None

#### execstats: add latch wait time into cumulative contention time

This patch adds a field `IsLatch` to the `ContentionEvent` protobuf struct that
indicates that a contention event was emitted by the latch table. The event is
added to the trace and used to increment the contention time metrics, but is
only added to the contention registry if
`register_latch_wait_contention_events` is set to true.

This commit also propagates the transaction ID into the latch manager and the
contention events listener. This information is used during a latch wait event
to distinguish between time waiting for a transaction's own requests
(e.g. for QueryIntent) and for real contention between different transactions.

Fixes #103713

Release note (ui change): The contention time metric now includes time spent
waiting to acquire latches in addition to time spent acquiring locks. This
applies to both the cumulative contention time visible in the db console and
the annotations of an `EXPLAIN ANALYZE` output.

DrewKimball committed:
execstats: add latch wait time into cumulative contention time

This patch adds a field `IsLatch` to the `ContentionEvent` protobuf struct that
indicates that a contention event was emitted by the latch table. The event is
added to the trace and used to increment the contention time metrics, but is
only added to the contention registry if
`register_latch_wait_contention_events` is set to true.

This commit also propagates the transaction ID into the latch manager and the
contention events listener. This information is used during a latch wait event
to distinguish between time waiting for a transaction's own requests
(e.g. for QueryIntent) and for real contention between different transactions.

Fixes #103713

Release note (ui change): The contention time metric now includes time spent
waiting to acquire latches in addition to time spent acquiring locks. This
applies to both the cumulative contention time visible in the db console and
the annotations of an `EXPLAIN ANALYZE` output.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/113649)
<!-- Reviewable:end -->

yuzefovich reviewed (commented):
Nice! This deserves some tests, but it seems like it'd be great to backport this.

> Reviewed 7 of 7 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/113649#-:-NiGS_-MEjcJMX-ahmJO:b-1n6nii)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball)

___
*[`-- commits` line 12 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/113649#-NiGSA3a3dW00WcEDuol):*
nit: this will also show up in EXPLAIN ANALYZE output.
___
*[`pkg/kv/kvserver/spanlatch/manager.go` line 599 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/113649#-NiGRgJz4zYIFGC4Lw3m:-NiGRgJz4zYIFGC4Lw3n:b59rlys) ([raw file](https://github.com/cockroachdb/cockroach/blob/a676b86f99e50f58ee93f3adef3b2f93f843a36a/pkg/kv/kvserver/spanlatch/manager.go#L599)):*
> ```Go
> 
> func maybeLogWaitTime(ctx context.Context, startTime time.Time) {
> 	waitTime := timeutil.Since(startTime)
> ```

nit: perhaps compute `waitTime` only inside the `if`.


<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Hi @DrewKimball, this looks great. Is there anything KV can do to help this proceed?

DrewKimball commented:
>Hi @DrewKimball, this looks great. Is there anything KV can do to help this proceed?

Sorry for the delay, I'll go ahead and fix this up with some tests today.

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/113649)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball)

___
*[`-- commits` line 12 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/113649#-NiGSA3a3dW00WcEDuol):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: this will also show up in EXPLAIN ANALYZE output.
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/spanlatch/manager.go` line 599 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/113649#-NiGRgJz4zYIFGC4Lw3m:-Nnv6xPn4J1P3QABzBkw:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a676b86f99e50f58ee93f3adef3b2f93f843a36a/pkg/kv/kvserver/spanlatch/manager.go#L599)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: perhaps compute `waitTime` only inside the `if`.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

DrewKimball commented:
@nvanbenschoten actually, I'm not so sure this works as-is, because there are sources of latch wait time that don't indicate contention. An example is mentioned [here](https://cockroachlabs.atlassian.net/browse/DOC-961) - QueryIntent requests that wait on a transaction's writes to finish. I don't know of other cases, but wouldn't be surprised if they exist. Maybe this means `spanLatch.Manager` is too low-level, because we need to know what txn is holding latches?

yuzefovich commented:
One thing to call out (which I was reminded of by [this](https://cockroachlabs.slack.com/archives/C063CP41TG9/p1713463101755419) slack thread) is we should check that latch contention events make their way to `crdb_internal.*_contention_events` vtables (on a quick glance it should just work).

kevin-v-ngo commented:
Any update on this issue @DrewKimball? Curious, could we target 24.2 (even better if we can backport) for this one?

Rooting out contention (including latch) symptoms continues to be a pain in several customer escalations.

DrewKimball commented:
>Any update on this issue @DrewKimball? Curious, could we target 24.2 (even better if we can backport) for this one?

>Rooting out contention (including latch) symptoms continues to be a pain in several customer escalations.

I can work on it this week - the remaining work is to keep track of the txn ID during latching so that we can tell between "self" and "contention" latch wait time (see a brief discussion [here](https://cockroachlabs.slack.com/archives/C0KB9Q03D/p1705959806724029)). @kevin-v-ngo do you have any thoughts on the metrics we want to surface here? I'm thinking of adding latch wait time involving other transactions directly to our `cumulative contention time` metric, but I can imagine we might also want to track something like `cumulative latch wait time` separately.

DrewKimball commented:
We also have the option to track "lock" vs "latch" contention time separately.

kevin-v-ngo commented:
> I can imagine we might also want to track something like cumulative latch wait time separately

Yes, I agree. I'd follow the same pattern as what we did for CPU (which only has SQL for now) and break it down.

1. One metric for latch
2. One metric for Lock
3. One metric for Latch+Lock (contentionTime as it is today in crdb_internal.statement_statistics but includes latch wait time)

This approach wouldn't alter the schema and precents potentially breaking customers using 'contentionTime' (it's only additive with two columns). And then in the UI, we can break it down, especially if there guidance on what a user would do when seeing one type of contention over another. 

Thoughts?

DrewKimball commented:
It seems difficult to write a good test for this, but I manually ran KV0 with `--zipfian` and collected a bundle. The bundle showed a bit less than a second waiting on a latch acquisition:

<img width="934" alt="Screenshot 2024-05-06 at 7 41 09 PM" src="https://github.com/cockroachdb/cockroach/assets/46886010/245a1b04-ae8e-4ea5-a8d6-58a6045428c5">

This was reflected in the structured stats event in the trace: 

```
  1002.647ms   1002.407ms            event:ComponentStats{ID: {16fb992d-281f-4d3f-9d9e-7f110b76a8a4 PROCESSOR 1 1 }, KV contention time: 989ms, execution time: 1s, sql cpu time: 2ms, batches output: 1, rows output: 1}
```

DrewKimball commented:
In the current iteration, I've plumbed the txn ID through the latch manager so that we can distinguish whether a latch wait was self-induced or due to contention. Now, only contention-related latch wait events are added to the cumulative contention time metric.

> Yes, I agree. I'd follow the same pattern as what we did for CPU (which only has SQL for now) and break it down.
> 
> One metric for latch
> One metric for Lock
> One metric for Latch+Lock (contentionTime as it is today in crdb_internal.statement_statistics but includes latch wait time)
> This approach wouldn't alter the schema and precents potentially breaking customers using 'contentionTime' (it's only additive with two columns). And then in the UI, we can break it down, especially if there guidance on what a user would do when seeing one type of contention over another.
> 
> Thoughts?

I think we should clarify now how we want to handle "self" vs "contention" vs "total" (self + contention) latch wait time. Should we just ignore self-induced latch waiting, and only track contention-related events in `One metric for latch`?

That being said, maybe we can defer that decision to another PR - this one will add latch wait time to the contention metric, and also adds enough tracking that we can easily add a metric self-induced or total latch wait time as well.

yuzefovich reviewed (commented):
Nice! LGTM.

Could you also manually double check this please

>  we should check that latch contention events make their way to `crdb_internal.*_contention_events` vtables (on a quick glance it should just work)

> Reviewed 3 of 3 files at r2, 12 of 17 files at r3, 7 of 7 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/113649#-:-NxIreuoARPS65w3-knl:bc4zgpn)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball)

___
*[`pkg/kv/kvserver/concurrency/concurrency_manager.go` line 374 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/113649#-NxIpxXBCmCXfQfQkSmv:-NxIpxXBCmCXfQfQkSmw:bej6bux) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb8b294d7d619dd517d9e11532dde36036cb4b87/pkg/kv/kvserver/concurrency/concurrency_manager.go#L374)):*
> ```Go
> 		// expire or to move to a finalized state.
> 		t := req.Requests[0].GetPushTxn()
> 		tBegin := timeutil.Now()
> ```

nit: not sure if this is worth micro-optimizing, but we could avoid the call to `timeutil.Now()` with something like:
```
sp := tracing.SpanFromContext(ctx)
if sp != nil {
  tBegin := timeutil.Now()
  defer func() { ... }
}
```
___
*[`pkg/sql/colfetcher/colbatch_scan.go` line 53 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/113649#-NxIr8TR5mJ9rPipO6jZ:-NxIr8TR5mJ9rPipO6j_:bkfin7a) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb8b294d7d619dd517d9e11532dde36036cb4b87/pkg/sql/colfetcher/colbatch_scan.go#L53)):*
> ```Go
> 	// execution, and it will be finished when closing the operator.
> 	tracingSpan              *tracing.Span
> 	contentionEventsListener execstats.ContentionEventsListener
> ```

nit: why make it a named field? IIUC we could keep it embedded, add a call to `Init`, and we won't need to add somewhat redundant `GetContentionTime` implementations.


<!-- Sent from Reviewable.io -->

kevin-v-ngo commented:
Thanks Drew! Several questions below😅

> Should we just ignore self-induced latch waiting, and only track contention-related events in One metric for latch?
What does self-induced latching waiting tell the user - refactor their transaction?

I assume this means the SQL *transaction* is 'conflicting' with itself correct? Or is the same statement (consisting of multiple kv requests) conflicting with itself?

Since we're adding latch contention events to _crdb_internal.*_contention_events_, users should be able to detect the blocking transaction, including itself correct? If so, I don't see why it's critical to further break down since the blocking id would be itself.

Adding @kevinkokomani @nicktrav @nvanbenschoten @yuzefovich to weigh in on this as well in case this additional metric (self-inducing wait time) would help in any scenarios/previous escalations

kevinkokomani commented:
Thanks @DrewKimball and @kevin-v-ngo. The only part I have a question about is:

> rather, I think it'd be most useful during escalations

If this is the case then I would still hope there is an easy way to identify the metric, perhaps through a method of advanced observability + an internal runbook. Makes sense that we don't want the metric readily accessible on the DB Console or end-user views because it is't indicative of any problems or any actions that the user should take. However if we want to potentially identify an issue that specifically is occurring during the write pipelining, e.g. maybe this "self-induced waiting time" is higher than we might expect, then tracking it somewhere + helping Support identify it easily with a runbook of some sort would be ideal.

kevin-v-ngo commented:
Thanks for the context @DrewKimball.

> I don't think "conflicting" is the right word for when this happens. Rather, there are times when a statement/txn has to wait for a request to finish before it can continue evaluation, and this is handled through the latch manager...So, this self-induced waiting time is expected during normal execution, and isn't in itself indicative of any problems. I also think it would be confusing to combine with contention-induced waiting time even if the previous point wasn't true, since the mitigations and causes are different.

Agree to not combine with contention-induced waiting time, especially if this metric would commonly not be indicative of a problem and the user action is distinct from transaction conflicts today.

I don't see any issues with capturing this as a _separate_ metric at least in our statement statistics table if this could be useful in escalations for L1/L2 - we may need it in the future for unknown unknowns. I'm not sure what the lift/work involved here but happy to track as a separate issue as well.

DrewKimball commented:
>One thing to call out (which I was reminded of by [this](https://cockroachlabs.slack.com/archives/C063CP41TG9/p1713463101755419) slack thread) is we should check that latch contention events make their way to crdb_internal.*_contention_events vtables (on a quick glance it should just work).

Actually, this didn't work before because I introduced a `LatchWaitEvent` struct, while the existing infrastructure searches the trace for `ContentionEvent`. I've reworked this draft to just always use `ContentionEvent` instead, and added an `IsLatch` field to that struct that changes the message printed in the trace. Latch wait events do appear to show up in the `contention_events` tables as expected.

DrewKimball commented:
>If this is the case then I would still hope there is an easy way to identify the metric, perhaps through a method of advanced observability + an internal runbook. Makes sense that we don't want the metric readily accessible on the DB Console or end-user views because it is't indicative of any problems or any actions that the user should take. However if we want to potentially identify an issue that specifically is occurring during the write pipelining, e.g. maybe this "self-induced waiting time" is higher than we might expect, then tracking it somewhere + helping Support identify it easily with a runbook of some sort would be ideal.

The latch wait events will still be visible in the trace as lines that look like this:
```
event:latch table conflict with 285e526b-1cbe-41d7-982b-bb043762410a on /Table/106/1/0/0 for 0.909s
```
They just won't be included in the `cumulative contention time` if it's a self-induced wait. Does that seem reasonable, or do we also want to add some additional observability?

>I don't see any issues with capturing this as a separate metric at least in our statement statistics table if this could be useful in escalations for L1/L2 - we may need it in the future for unknown unknowns. I'm not sure what the lift/work involved here but happy to track as a separate issue as well.

Issue: https://github.com/cockroachdb/cockroach/issues/124068

yuzefovich reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Would be good to get an approval from KV too, cc @nvanbenschoten 

> Reviewed 12 of 14 files at r5, 2 of 2 files at r6, 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/113649#-:-NxndYhT8IWtmBOt8Wcs:b-rev874)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball, @nkodali, and @rytaft)

___
*[`-- commits` line 4 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/113649#-Nxnau6oCP-A1fDWvb_p):*
nit: the commit message and the PR description now need an update.


<!-- Sent from Reviewable.io -->

DrewKimball committed:
sql: add setting to enable registering latch wait contention events

This commit adds a session setting (default off) that will cause contention
events emitted by the latch table to be included in the `*_contention_events`
virtual tables. This behavior is hidden behind a setting because latch wait
events are more common than events emitted by the lock table. Tracking them
in the virtual tables could increase system load and flood the statement
insights with uninteresting events. For now, this setting does nothing, but
the following commit will use it.

Informs #103713

Release note: None

DrewKimball commented:
After some discussion with @yuzefovich, I've added a default-off session setting that determines whether these latch-wait contention events are added to the contention registry. My reasoning is that these events are likely much more common than other sources, and we don't want to add significant overhead to the system or flood the statement insights with uninteresting events.

The `cumulative contention time` metric in the db console and EXPLAIN ANALYZE diagram annotations will still reflect latch wait time, as will the `ContendedQueriesCount` and `CumulativeContentionNanos` metrics.

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/113649)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @nkodali, @rytaft, and @yuzefovich)

___
*[`-- commits` line 4 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/113649#-Nxnau6oCP-A1fDWvb_p):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: the commit message and the PR description now need an update.
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/concurrency/concurrency_manager.go` line 374 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/113649#-NxIpxXBCmCXfQfQkSmv:-NxT6oh-4l4CvGbFqHMM:b-fznogt) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb8b294d7d619dd517d9e11532dde36036cb4b87/pkg/kv/kvserver/concurrency/concurrency_manager.go#L374)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: not sure if this is worth micro-optimizing, but we could avoid the call to `timeutil.Now()` with something like:
```
sp := tracing.SpanFromContext(ctx)
if sp != nil {
  tBegin := timeutil.Now()
  defer func() { ... }
}
```
</blockquote></details>

Good idea, done.
___
*[`pkg/sql/colfetcher/colbatch_scan.go` line 53 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/113649#-NxIr8TR5mJ9rPipO6jZ:-NxT79eR89LIMYBuGOzp:b-h3wmqt) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb8b294d7d619dd517d9e11532dde36036cb4b87/pkg/sql/colfetcher/colbatch_scan.go#L53)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: why make it a named field? IIUC we could keep it embedded, add a call to `Init`, and we won't need to add somewhat redundant `GetContentionTime` implementations.
</blockquote></details>

I didn't think of that, good call.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (approved):
> Reviewed 24 of 24 files at r8, 26 of 26 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/113649)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball, @nkodali, and @rytaft)

___
*[`pkg/sql/crdb_internal_test.go` line 1200 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/113649#-Nxx1WwgEquizBiGZz3H:-Nxx1WwgEquizBiGZz3I:b6hc0rh) ([raw file](https://github.com/cockroachdb/cockroach/blob/c150a9d1e8b5602d93495e6888deb4740207c938/pkg/sql/crdb_internal_test.go#L1200)):*
> ```Go
> 
> 				if !strings.Contains(indexName, expectedIndex) {
> 					fmt.Println(indexName)
> ```

nit: seems like a leftover?


<!-- Sent from Reviewable.io -->

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> +1 on getting someone on KV to take a look at this one.

> Reviewed 24 of 24 files at r8, 26 of 26 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/113649#-:-Nxy3vImBz4vN0mzT5Ch:b-77lxl5)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball, @nkodali, and @rytaft)



<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/113649)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @nkodali, @rytaft, and @yuzefovich)

___
*[`pkg/sql/crdb_internal_test.go` line 1200 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/113649#-Nxx1WwgEquizBiGZz3H:-NyR9_8Y71unxw1NtNWz:b-k614gh) ([raw file](https://github.com/cockroachdb/cockroach/blob/c150a9d1e8b5602d93495e6888deb4740207c938/pkg/sql/crdb_internal_test.go#L1200)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: seems like a leftover?
</blockquote></details>

Yes, thanks.


<!-- Sent from Reviewable.io -->

nvb reviewed (dismissed):
> Reviewed 32 of 32 files at r10, 26 of 26 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/113649)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @DrewKimball, @nkodali, @rytaft, and @yuzefovich)

___
*[`pkg/kv/kvpb/api.go` line 2363 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/113649#-NyRN2492Hi2Zi5EB8rE:-NyRN2492Hi2Zi5EB8rF:b-1kq14z) ([raw file](https://github.com/cockroachdb/cockroach/blob/904f495c89365943ef87b7dd3f8cdf73c18fdfd4/pkg/kv/kvpb/api.go#L2363)):*
> ```Go
> 	prefix := redact.SafeString("conflicted")
> 	if c.IsLatch {
> 		prefix = "latch table conflict"
> ```

We have never really called anything the "latch table". Let's just say `prefix = "latch conflict"`.
___
*[`pkg/kv/kvpb/api.proto` line 3686 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/113649#-NyRNHiD6bQb-HAOJAfy:-NyRNHiD6bQb-HAOJAfz:b5fxd96) ([raw file](https://github.com/cockroachdb/cockroach/blob/904f495c89365943ef87b7dd3f8cdf73c18fdfd4/pkg/kv/kvpb/api.proto#L3686)):*
> ```proto
>   // IsLatch, if true, indicates that the contention event was due to waiting to
>   // acquire a latch.
>   bool is_latch = 5;
> ```

`bool is_latch = 4`?
___
*[`pkg/kv/kvserver/concurrency/concurrency_manager.go` line 375 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/113649#-NyRjEtE3_cteduG8T7g:-NyRjEtE3_cteduG8T7h:b-ok8vw0) ([raw file](https://github.com/cockroachdb/cockroach/blob/904f495c89365943ef87b7dd3f8cdf73c18fdfd4/pkg/kv/kvserver/concurrency/concurrency_manager.go#L375)):*
> ```Go
> 		// expire or to move to a finalized state.
> 		t := req.Requests[0].GetPushTxn()
> 		sp := tracing.SpanFromContext(ctx)
> ```

I'm confused about this new logic. Why is the time spent in `MaybeWaitForPush` considered a latch? Isn't this time already accounted for in the `lockTableWaiter`?
___
*[`pkg/kv/kvserver/spanlatch/manager.go` line 163 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/113649#-NyRiJlF7LAzVgF8ut7s:-NyRiJlF7LAzVgF8ut7t:b-r2sf1) ([raw file](https://github.com/cockroachdb/cockroach/blob/904f495c89365943ef87b7dd3f8cdf73c18fdfd4/pkg/kv/kvserver/spanlatch/manager.go#L163)):*
> ```Go
> 	snap        *snapshot
> 	acquireTime int64
> 	txn         *enginepb.TxnMeta
> ```

We recently (ba13697aac5a020bfcc6d0ffc10912eacb8b5d2d) added the `baFmt` field here with an intention to avoid a direct dependency on `*roachpb.BatchRequest`. If we're now going to be adding the `Txn` from the `BatchRequest.Header`, we should just give up and place the entire `*roachpb.BatchRequest` on the `Guard`.

So concretely, wherever we're now passing `baFmt redact.SafeFormatter, txn *roachpb.Transaction`, let's just pass `ba *roachpb.BatchRequest`.

An added benefit of this change will be that it shrinks `Guard`'s memory footprint instead of growing it,
___
*[`pkg/kv/kvserver/concurrency/testdata/concurrency_manager/basic` line 152 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/113649#-NyRkvfv1s58IY2fI8r4:-NyRkvfv1s58IY2fI8r5:b-u0qc8s) ([raw file](https://github.com/cockroachdb/cockroach/blob/904f495c89365943ef87b7dd3f8cdf73c18fdfd4/pkg/kv/kvserver/concurrency/testdata/concurrency_manager/basic#L152)):*
> ```
> ----
> write count: 1
>  read count: 1
> ```

We can add the following here so that we get some test coverage of the conflict duration:
```
debug-advance-clock ts=123
----
```

Feel free to add then elsewhere in these `testdata/concurrency_manager` tests if there's somewhere else that's also appropriate.
___
*[`pkg/kv/kvserver/concurrency/lock_table_test.go` line 1294 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/113649#-NyRlHuu0Jn8aNrhvQHO:-NyRlHuu0Jn8aNrhvQHP:byasgoc) ([raw file](https://github.com/cockroachdb/cockroach/blob/904f495c89365943ef87b7dd3f8cdf73c18fdfd4/pkg/kv/kvserver/concurrency/lock_table_test.go#L1294)):*
> ```Go
> 		concurrency:  concurrency,
> 	}
> 	ex.lm = spanlatch.Make(
> ```

nit: we can construct this above like `lt` so that we can assign it in the constructor and revert back to `return &workloadExecutor`. Just like you're doing down below in `BenchmarkLockTable`.


<!-- Sent from Reviewable.io -->

kevin-v-ngo commented:
Any update on this @DrewKimball? Contention continues to be a top issue in escalations we rule out/troublshoot.

To confirm, could we have 3 separate metrics in statement stats? Seems like this is tracking only the cumulative.

1. One metric for Latch
2. One metric for Lock
3. One cumulative metric for Latch+Lock (contentionTime as it is today in crdb_internal.statement_statistics but includes latch wait time)

https://github.com/cockroachdb/cockroach/pull/113649#issuecomment-2097381131

DrewKimball committed:
sql: add lock and latch wait time metrics

This commit plumbs cumulative lock and latch wait durations from the
contention events listener up to `QueryLevelStats`. The metrics are
displayed for `EXPLAIN ANALYZE` variants. Surfacing these metrics in
the DB console, statement/txn insights etc. is left for future work.

Informs #103713

Release note (ui change): The lock and latch wait time components of
a query's cumulative contention time are now tracked separately and
surfaced as annotations in `EXPLAIN ANALYZE` output.

DrewKimball reviewed (commented):
Belated TFTRs!

>To confirm, could we have 3 separate metrics in statement stats? Seems like this is tracking only the cumulative.
>
>1.  One metric for Latch
>2.  One metric for Lock
>3.  One cumulative metric for Latch+Lock (contentionTime as it is today in crdb\_internal.statement\_statistics but includes latch wait time)

I've added a commit that surfaces lock and latch wait time as separate `EXPLAIN ANALYZE` annotations. I've left propagating these to the DB console as future work, for now.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/113649#-:-OKhxehq3SMOIk96VIfD:b-avn8ni)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @nkodali, @nvanbenschoten, @rytaft, and @yuzefovich)

___
*[`pkg/kv/kvpb/api.go` line 2363 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/113649#-NyRN2492Hi2Zi5EB8rE:-OKcsrC88guGnSjLLWzd:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/904f495c89365943ef87b7dd3f8cdf73c18fdfd4/pkg/kv/kvpb/api.go#L2363)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

We have never really called anything the "latch table". Let's just say `prefix = "latch conflict"`.
</blockquote></details>

Done.
___
*[`pkg/kv/kvpb/api.proto` line 3686 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/113649#-NyRNHiD6bQb-HAOJAfy:-NySdDQp8hff_9h9VHLQ:b-4tqtz9) ([raw file](https://github.com/cockroachdb/cockroach/blob/904f495c89365943ef87b7dd3f8cdf73c18fdfd4/pkg/kv/kvpb/api.proto#L3686)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

`bool is_latch = 4`?
</blockquote></details>

Good catch, Done.
___
*[`pkg/kv/kvserver/concurrency/concurrency_manager.go` line 375 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/113649#-NyRjEtE3_cteduG8T7g:-NySe1md85MSuGdUsDNS:biuwmhv) ([raw file](https://github.com/cockroachdb/cockroach/blob/904f495c89365943ef87b7dd3f8cdf73c18fdfd4/pkg/kv/kvserver/concurrency/concurrency_manager.go#L375)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I'm confused about this new logic. Why is the time spent in `MaybeWaitForPush` considered a latch? Isn't this time already accounted for in the `lockTableWaiter`?
</blockquote></details>

This goes through the `txnWaitQueue` though, not the `lockTableWaiter`, right? With the actual waiting happening in [Queue.waitForPush()](https://github.com/cockroachdb/cockroach/blob/b589a84c9c3f6450de1dd674ba7f9b0fe274c17c/pkg/kv/kvserver/txnwait/queue.go#L566) on the channel created [here](https://github.com/cockroachdb/cockroach/blob/b589a84c9c3f6450de1dd674ba7f9b0fe274c17c/pkg/kv/kvserver/txnwait/queue.go#L522). Maybe it's wrong to classify it as latch-wait time, though... I'll remove it.
___
*[`pkg/kv/kvserver/spanlatch/manager.go` line 163 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/113649#-NyRiJlF7LAzVgF8ut7s:-NySjLf_7QfMgHW5MCIj:b-p5qbgk) ([raw file](https://github.com/cockroachdb/cockroach/blob/904f495c89365943ef87b7dd3f8cdf73c18fdfd4/pkg/kv/kvserver/spanlatch/manager.go#L163)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

We recently (ba13697aac5a020bfcc6d0ffc10912eacb8b5d2d) added the `baFmt` field here with an intention to avoid a direct dependency on `*roachpb.BatchRequest`. If we're now going to be adding the `Txn` from the `BatchRequest.Header`, we should just give up and place the entire `*roachpb.BatchRequest` on the `Guard`.

So concretely, wherever we're now passing `baFmt redact.SafeFormatter, txn *roachpb.Transaction`, let's just pass `ba *roachpb.BatchRequest`.

An added benefit of this change will be that it shrinks `Guard`'s memory footprint instead of growing it,
</blockquote></details>

Nice idea, Done. Most of the fields on `concurrency.Request` are also coming from the batch - should I refactor that as well?
___
*[`pkg/kv/kvserver/concurrency/testdata/concurrency_manager/basic` line 152 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/113649#-NyRkvfv1s58IY2fI8r4:-OKcxw_YBW_zLt6QGgbh:b-574kmi) ([raw file](https://github.com/cockroachdb/cockroach/blob/904f495c89365943ef87b7dd3f8cdf73c18fdfd4/pkg/kv/kvserver/concurrency/testdata/concurrency_manager/basic#L152)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

We can add the following here so that we get some test coverage of the conflict duration:
```
debug-advance-clock ts=123
----
```

Feel free to add then elsewhere in these `testdata/concurrency_manager` tests if there's somewhere else that's also appropriate.
</blockquote></details>

Nice, Done.
___
*[`pkg/kv/kvserver/concurrency/lock_table_test.go` line 1294 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/113649#-NyRlHuu0Jn8aNrhvQHO:-OKcsBaP7THZNr_jYMMo:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/904f495c89365943ef87b7dd3f8cdf73c18fdfd4/pkg/kv/kvserver/concurrency/lock_table_test.go#L1294)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: we can construct this above like `lt` so that we can assign it in the constructor and revert back to `return &workloadExecutor`. Just like you're doing down below in `BenchmarkLockTable`.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (approved):
> Reviewed 32 of 32 files at r12, 28 of 28 files at r13, 29 of 29 files at r14, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/113649)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @nkodali, @nvanbenschoten, and @rytaft)

___
*[`pkg/sql/execstats/traceanalyzer.go` line 391 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/113649#-OKi2EpX5U_okc5MepJd:-OKi2EpX5U_okc5MepJe:b-9e4l2i) ([raw file](https://github.com/cockroachdb/cockroach/blob/86faf5050acf5c53ad3b58fe7047e4867e2873d7/pkg/sql/execstats/traceanalyzer.go#L391)):*
> ```Go
> 				return
> 			}
> 			if ev.IsLatch && !evalCtx.SessionData().RegisterLatchWaitContentionEvents {
> ```

nit: I think it'd be cleaner if we skipped this event in `recordStatementSummary` and not here. The name `getAllContentionEvents` and the comment on the method suggest that all contention events should be returned, and now that wouldn't be the case.


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/113649)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @nkodali, @nvanbenschoten, and @rytaft)

___
*[`pkg/sql/execstats/traceanalyzer.go` line 391 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/113649#-OKi2EpX5U_okc5MepJd:-OKjLF2HBCx9JndC7T9H:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/86faf5050acf5c53ad3b58fe7047e4867e2873d7/pkg/sql/execstats/traceanalyzer.go#L391)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: I think it'd be cleaner if we skipped this event in `recordStatementSummary` and not here. The name `getAllContentionEvents` and the comment on the method suggest that all contention events should be returned, and now that wouldn't be the case.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

DrewKimball commented:
bors r+

yuzefovich commented:
(Dismissed stale review from Nathan.)

bors r+
